designs/data_serialization_design.md

```diff
@@ -0,0 +1,189 @@
+# FAIR Data Serialization Design
+## Overview
+
+This document outlines the design for serializing FAIR’s core data structures, which is essential for enabling FAIR as a distributed system. It also covers how serialized data is expected to be consumed and synchronized across hosts. 
+
+This document doesn't capture details about how the serialized states from other hosts are merged/considered in the decision making.
+
+## Glossary
+
+What's a Bucket? \
+The most basic data-structure in FAIR implementatin is a bucket - this contains a probability and a last-seen time. The probability is decayed (or) improved based on some factors. For the life-time of the Structure, a client - identified by the client-ID, consistently lands on a few buckets (based on seed). The final decision is made on top of the bucket probabilities.
+
+
+## Requirements
+### Functional
+
+- **Complete**: Serialize the full data structure, including:
+    - Multi-level bucket data (probabilities, timestamps)
+    -  Configuration parameters
+    - MurmurHash seed
+    - Metadata (ID, flags)
+
+- **Portable**: Work across platforms, Go versions, and FAIR versions, with proper versioning.
+- **Compact**: Add minimal overhead for efficient transfer.
+- **Future-proof**: Support schema evolution as new parameters are added.
+
+### Non-Functional
+- **Performance**: Fast serialization/deserialization.
+- **Reliability**: Strong validation and error handling.
+- **Maintainability**: Clear versioning and schema upgrade path.
+- **Compatibility**: Backward and forward compatibility.
+
+## Serialization Format Selection
+
+Rejected options: JSON, XML, and other text formats — too verbose. \
+Custom serialization: Considered, but unnecessary given current requirements. Can revisit if protobuf overhead becomes a concern. \
+Chosen option: Protocol Buffers (protobuf), due to:
+- Binary format
+- Native support in gRPC
+- Strong typing and schema evolution
+- Built-in JSON support (for debugging)
+- Schema Design
+
+The schema has two broad parts:
+1)  Config (TrackerCfg) — defined by customers, drives tracker creation.
+2) Runtime Data — driven by traffic and execution:
+   - Consts: Initialized once, updated periodically (e.g., hash seed).
+   - State: Updated frequently on the hot path (bucket data).
+
+Accessing `State` requires the matching `Config` and `Consts`. Multiple tracker configs may exist simultaneously (multi-tenant).
```

Comment:
Not sure what you mean by multi-tenant here?

Reply:
There is another [place](https://github.com/satmihir/fair/pull/44/files#r2366634468) where we are discussing this. I sense there is a disconnect in our lingo :)

Taking a different stab at this -  I am thinking, we will introduce a wrapper on top of the `FairnessTracker`.\
We expose an RPC server, and client invokes it with `tenantID` and `clientID`. Each `tenantID` has its own FairnessTracker object.

Reply:
I still don't understand what you mean by a tenant :)

In my mind, every tracker is a use case - one for calling your database, one for calling metadata service etc. If doing server side throttling, maybe one per http path scheme handler etc. So a single service has n trackers. Is this what you mean by a tenant? I wouldn't use that term for it. Perhaps a tracker ID or something?

Also, this Tracker will eventually go into an http client and when different hosts collaborate, they sync on the state of various trackers individually (perhaps batch it but we'll see. I don't anticipate hundreds of these per service - more like 2-3.